library/autodiff/examples/neohookean.rs

```diff
@@ -0,0 +1,363 @@
+    pub fn from_lame(lambda: f64, mu: f64) -> Self {
+        Self { lambda, mu }
+    }
+    // https://en.wikipedia.org/wiki/Lam%C3%A9_parameters
+    pub fn from_youngs(E: f64, nu: f64) -> Self {
+        Self {
+            lambda: E * nu / ((1.0 + nu) * (1.0 - 2.0 * nu)),
+            mu: E / (2.0 * (1.0 + nu)),
+        }
+    }
+}
+
+// Helmholtz free energy density
+#[autodiff(d_psi, Reverse, Active)]
+fn psi(#[dup] e: &KM, nh: &NH) -> f64 {
+    let mu = nh.mu;
+    let lambda = nh.lambda;
+    let J = e.cauchy_green().det().sqrt();
+    0.25 * lambda * (J * J - 1.0 - 2.0 * J.ln()) // convex
+      + mu * (e.trace() - J.ln())
+}
+
+// Kinchhoff stress (current configuration)
+#[autodiff(d_stress, Forward, DuplicatedNoNeed)]
+fn stress(#[dup] e: &KM, nh: &NH) -> KM {
+    let mut dpsi_de = KM::zero();
+    d_psi(&e, &mut dpsi_de, &nh, 1.0);
+    let b = e.cauchy_green();
+    dpsi_de * b
+}
+
+pub mod analytic {
+    use super::*;
+    pub fn stress(e: &KM, nh: &NH) -> KM {
+        let lambda = nh.lambda;
+        let mu = nh.mu;
+        let b = e.cauchy_green();
+        let J = b.det().sqrt();
+        let I = KM::identity();
+        0.5 * lambda * (J * J - 1.0) * I + 2.0 * mu * e
+    }
+
+    pub fn d_stress(e: &KM, ddudx: &Mat3x3, nh: &NH) -> KM {
+        let lambda = nh.lambda;
+        let mu = nh.mu;
+        let b = e.cauchy_green();
+        let J = b.det().sqrt();
+        let tau = stress(e, nh).to_matrix();
+        let deps = KM::epsilon(ddudx);
+        let I = KM::identity();
+        let FdSFt = lambda * J * J * deps.trace() * I + (2.0 * mu - lambda * (J * J - 1.0)) * deps;
+        2.0 * KM::from_matrix(matmul(&ddudx, false, &tau, false)) + FdSFt
+    }
+}
+
+fn main() {
+    let nh = NH::from_lame(2.0, 1.0);
+    let e = KM::from_voigt(&[0.2, -0.1, -0.12, 0.3, 0.2, 0.15]);
+    let energy = psi(&e, &nh);
+    println!("energy: {energy}");
+
+    let mut dpsi_de = KM::zero();
+    d_psi(&e, &mut dpsi_de, &nh, 1.0);
+    println!("dpsi_de: {dpsi_de:?}");
+    let stress = stress(&e, &nh);
+    println!("enzyme stress  : {stress:?}");
+    let stress = analytic::stress(&e, &nh);
+    println!("analytic stress: {stress:?}");
+
+    let ddudx = [[1.0, 1.1, 1.2], [1.3, 1.4, 1.5], [1.6, 1.7, 1.8]];
+    let deps = KM::epsilon(&ddudx);
+    if false {
+        // rustc: /home/jeka2967/src/rust/src/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:8581: llvm::ScalarEvolution::ExitLimit llvm::ScalarEvolution::computeExitLimitFromSingleExitSwitch(const llvm::Loop*, llvm::SwitchInst*, llvm::BasicBlock*, bool): Assertion `L->contains(Switch->getDefaultDest()) && "Default case must not exit the loop!"' failed.
+        let de = 2.0 * KM::from_matrix(matmul(&ddudx, false, &e.to_matrix(), false)) + deps;
+        let dtau = d_stress(&e, &de, &nh);
+        println!("enzyme dstress  : {dtau:?}");
+    }
+    let dtau = analytic::d_stress(&e, &ddudx, &nh);
+    println!("enzyme dstress  : {dtau:?}");
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I'm not sure if you're currently running `cargo +enzyme test --examples`, but I find it helpful to have these unit tests to supplement the run output.

Reply:
Noticed that you added those recently, looks like a nice improvement to me, thanks.